system/lib/libc/musl/src/thread/pthread_mutex_lock.c

```diff
@@ -2,9 +2,12 @@
 
 int __pthread_mutex_lock(pthread_mutex_t *m)
 {
+#if !defined(__EMSCRIPTEN__) || defined(NDEBUG)
+	/* XXX EMSCRIPTEN always take the slow path in debug builds so we can trap rather than deadlock */
 	if ((m->_m_type&15) == PTHREAD_MUTEX_NORMAL
 	    && !a_cas(&m->_m_lock, 0, EBUSY))
 		return 0;
+#endif
```

Comment:
Should upstream do the same for all platforms? I'm just trying to understand if this is a general improvement or if it works around a specific problem in wasm.

Reply:
Upstream musl doesn't really have any/many assertions AFAICT, so its not something they tend to consider I suppose?

Reply:
Indeed, it looks like upstream musl only has assertion in 2 very specific places:

```
src/regex/regcomp.c
src/malloc/mallocng/..
```

No other assertions in whole of libc.

Reply:
I see, thanks. Makes sense.

Reply:
I was referring to "Upstream musl doesn't really have any/many assertions AFAICT" in general, the question was whether llvm-libc tends to do assertions for debug builds more than musl does.

Reply:
Apparently they (llvm libc folks) are open to adding this kind of asserts.

Reply:
That sounds promising, being able to upstream these things makes your argument about the switch from musl to llvm-libc even more compelling.